File: ga/selector.go
package main

import (
	"math/rand"
	"sort"
)

type Selector interface {
	Select(population Population) Individual
}

type Tournament struct {
	tournamentSize int
}

func (selector Tournament) Select(population Population) Individual {
	var competitors []Individual
	for i := 0; i < selector.tournamentSize; i++ {
		randomNumber := rand.Intn(population.size) // range?
		competitor := population.individuals[randomNumber]
		competitors = append(competitors, competitor)
	}
	winner := competitors[0]
	for _, competitor := range competitors {
		if competitor.fitness < winner.fitness {
			continue
		}
		winner = competitor
	}
	return winner
}

type Roulette struct{}

func (Roulette) Select(population Population) Individual {
	//if population.totalFitness == 0 {
	//	randomNumber := rand.Intn(len(population.individuals))
	//	return population.individuals[randomNumber]
	//}
	randomNumber := rand.Float64()
	probabilitySum := 0.0
	for _, individual := range population.individuals {
		probabilitySum += individual.fitness / population.totalFitness
		if randomNumber < probabilitySum {
			return individual
		}
	}
	panic("Couldn't select an individual")
}

type Rank struct {
	selectionPressure float64
}

func (selector Rank) Select(population Population) Individual {
	sort.Slice(population.individuals, func(i, j int) bool {
		return population.individuals[i].fitness > population.individuals[j].fitness
	})
	randomNumber := rand.Float64()
	probabilitySum := 0.0
	for i, individual := range population.individuals {
		probabilitySum += (1 / float64(population.size)) * (selector.selectionPressure - (2*selector.selectionPressure-2)*float64(i)/(float64(population.size)-1))
		if randomNumber < probabilitySum {
			return individual
		}
	}
	panic("Couldn't select an individual")
}

File: ui/src/lib/ts/ga/selector.ts
import type { Individual } from "./individual";
import type { Population } from "./population";
// import * as _ from "lodash";

export interface Selector {
  select(population: Population): Individual;
}

// export class Tournament implements Selector {
//   tournamentSize: number;

//   constructor(tournamentSize: number) {
//     this.tournamentSize = tournamentSize;
//   }

//   select(population: Population): Individual {
//     const competitors: Individual[] = _.sampleSize(population.individuals, this.tournamentSize);
//     const winner = competitors.reduce((currentWinner, competitor) => competitor.fitness >= currentWinner.fitness ? competitor : currentWinner);
//     return winner;
//   }
// }

export class CustomSelector implements Selector {
  select(population: Population): Individual {
    throw new Error("Custom select not implemented");
  }
}

